### PR TITLE
Additional special functions in Microsoft.Quantum.Math.

### PR DESCRIPTION
**Pull request by cgranade**
Please see #448 for feature request, #457 for API design approval, and #440 for initial contribution (thanks, @TheMagicNacho!).

### CODE REVIEW

**Comment by cgranade** [comment]
Looks like the above was a transient build failure; re-running.

File: Standard/src/Math/Functions.qs [hunk 1]
```diff
@@ -655,6 +656,274 @@ namespace Microsoft.Quantum.Math {
...
+            fail "Largest factorial an Int can hold is 20!. Use FactorialL or ApproximateFactorial.";
+        } else {
+            set an = n;
+        }
+
+        for i in  1 .. an {
+            set x *= i;
+        }
+
+       return x;
```
Comment (msoeken) [inline Standard/src/Math/Functions.qs:696]:
```suggestion
        return x;
```

File: Standard/src/Math/Functions.qs [hunk 2]
```diff
@@ -655,6 +656,274 @@ namespace Microsoft.Quantum.Math {
         }
     }
 
-}
+
+    /// # Summary
+    /// Returns the factorial of a given number.
+    ///
+    /// # Description
+    /// Returns the factorial of a given integer $n$, where $|n| <= 20$.
```
Comment (msoeken) [inline Standard/src/Math/Functions.qs:664]:
```suggestion
    /// Returns the factorial of a given nonnegative integer $n$, where $n \le 20$.
```

File: Standard/src/Math/Functions.qs [hunk 3]
```diff
@@ -655,6 +656,274 @@ namespace Microsoft.Quantum.Math {
...
+
+       return x;
+    }
 
 
+    /// # Summary
+    /// Returns an approximate factorial of a given number.
+    ///
+    /// # Description
+    /// Returns the factorial as 'Double', given an input of $n$ as a 'Double'.
```
Comment (msoeken) [inline Standard/src/Math/Functions.qs:704]:
```suggestion
    /// Returns the factorial as `Double`, given an input of $n$ as a `Double`.
```

File: Standard/src/Math/Functions.qs [hunk 4]
```diff
@@ -655,6 +656,274 @@ namespace Microsoft.Quantum.Math {
...
+    /// # Remarks
+    /// For inputs greater than 20, please use @"Microsoft.Quantum.Math.FactorialL".
+    ///
+    /// # See Also
+    /// - Microsoft.Quantum.Math.FactorialL
+    function FactorialI(n : Int) : Int {
+        mutable an = 1;
+        mutable x = 1;
+
+        if n < 0 {
```
Comment (msoeken) [inline Standard/src/Math/Functions.qs:682]: Would it make sense to extract clauses with a `fail` statement into `Fact` statements and then keep the control logic simpler taking into account new assumptions from the `Fact` statements?

File: Standard/src/Math/Functions.qs [hunk 5]
```diff
@@ -655,6 +656,274 @@ namespace Microsoft.Quantum.Math {
...
+    ///
+    /// # Output
+    /// The approximate factorial of `n`.
+    ///
+    /// # See Also
+    /// - Microsoft.Quantum.Math.FactorialI
+    /// - Microsoft.Quantum.Math.FactorialL
+    function ApproximateFactorial(n : Int) : Double {
+        if n < 0 {
+            fail "The factorial is not defined for negative inputs.";
```
Comment (msoeken) [inline Standard/src/Math/Functions.qs:723]: Same comment as above regarding `Fact` statements.

File: Standard/src/Math/Functions.qs [hunk 6]
```diff
@@ -655,6 +656,274 @@ namespace Microsoft.Quantum.Math {
...
+    /// $7!! = 7 \times 5 \times 3 \times 1$.
+    ///
+    /// # See Also
+    /// - Microsoft.Quantum.Math.ApproximateFactorial
+    /// - Microsoft.Quantum.Math.FactorialI
+    /// - Microsoft.Quantum.Math.FactorialL
+    function DoubleFactorialL(n : Int) : BigInt {
+        if n < 0 {
+            fail "The double factorial is not defined for negative inputs.";
+        }
```
Comment (msoeken) [inline Standard/src/Math/Functions.qs:759]:
```suggestion
        Fact(n >= 0, "The double factorial is not defined for negative inputs.");
```

File: Standard/src/Math/Functions.qs [hunk 7]
```diff
@@ -655,6 +656,274 @@ namespace Microsoft.Quantum.Math {
...
+    /// # Remarks
+    /// The double factorial $n!!$ of $n$ is defined as
+    /// $n \times (n - 2) \times \cdots \times k$, where $k \in {1, 2}$. For example,
+    /// $7!! = 7 \times 5 \times 3 \times 1$.
+    ///
+    /// # See Also
+    /// - Microsoft.Quantum.Math.ApproximateFactorial
+    /// - Microsoft.Quantum.Math.FactorialI
+    /// - Microsoft.Quantum.Math.FactorialL
+    function DoubleFactorialL(n : Int) : BigInt {
```
Comment (msoeken) [inline Standard/src/Math/Functions.qs:756]: This function was not part of the API review and I suggest to mark it `internal` or add it to the API issue.

File: Standard/src/Math/Functions.qs [hunk 8]
```diff
@@ -655,6 +656,274 @@ namespace Microsoft.Quantum.Math {
...
+        // Here, we use the approximation described in Numerical Recipes in C.
+        let coefficients = [
+            57.1562356658629235, -59.5979603554754912,
+            14.1360979747417471, -0.491913816097620199, .339946499848118887e-4,
+            .465236289270485756e-4, -.983744753048795646e-4, .158088703224912494e-3,
+            -.210264441724104883e-3, .217439618115212643e-3, -.164318106536763890e-3,
+            .844182239838527433e-4, -.261908384015814087e-4, .368991826595316234e-5
+        ];
+
+        if x <= 0.0 { fail "Γ(x) not defined for x <= 0."; }
```
Comment (msoeken) [inline Standard/src/Math/Functions.qs:853]:
```suggestion
        Fact(x > 0.0, "Γ(x) not defined for x <= 0.");
```

**Review by msoeken — APPROVED** [review]
Looks good to me. Thanks @cgranade.

File: Standard/src/Math/Functions.qs (hunk 7 above)
Comment (cgranade) [reply Standard/src/Math/Functions.qs:756]: Ah, thank you for catching that... needed to support so yes, will mark as internal.

File: Standard/src/Math/Functions.qs (hunk 4 above)
Comment (cgranade) [reply Standard/src/Math/Functions.qs:682]: Good idea, I'll go on and do that. Thanks!

File: Standard/src/Math/Functions.qs (hunk 5 above)
Comment (cgranade) [reply Standard/src/Math/Functions.qs:723]: Done.

File: Standard/src/Math/Functions.qs (hunk 8 above)
Comment (cgranade) [reply Standard/src/Math/Functions.qs:851]: Fixed.

**Comment by TheMagicNacho** [comment]
=D This is great to see!

**Pull request merged**